ddspipe_participants/src/cpp/reader/dds/CommonReader.cpp

```diff
@@ -128,41 +128,42 @@ utils::ReturnCode CommonReader::take_nts_(
         return utils::ReturnCode::RETCODE_NO_DATA;
     }
 
-    RtpsPayloadData* rtps_data;
+    std::unique_ptr<RtpsPayloadData> rtps_data;
     fastdds::dds::SampleInfo info;
 
-    // Loop for read messages until we receive one that does not come from same participant
-    // NOTE: not reading local messages must be done in this loop because:
-    // 1. there is no way in DDS to ignore this reader from writer as there were in RTPS
-    // 2. ignore_local_endpoints would be override by xml configuration
-    while (true)
+    do
     {
-        // Ensure that the previous Payload gets destroyed to avoid memory leaks.
-        rtps_data = new core::types::RtpsPayloadData();
-        data.reset(rtps_data);
+        // The reset method does two things:
+        //      1. It releases the previous object stored in data.
+        //      2. It transfers the ownership of rtps_data to data.
+        rtps_data.reset(new RtpsPayloadData());
 
-        auto ret = reader_->take_next_sample(rtps_data, &info);
+        auto ret = reader_->take_next_sample(rtps_data.get(), &info);
 
-        // Save the payload owner so that the memory is freed correctly.
+        // If the payload owner is not set, rtps_data won't release the payload on destruction
         rtps_data->payload_owner = payload_pool_.get();
 
-        // If error reading data
         if (!ret)
         {
+            // There has been an error taking the data. Exit.
             return ret;
         }
-
-        // Check if the sample is acceptable
-        if (should_accept_sample_(info))
-        {
-            break;
-        }
-    }
+    } while (!should_accept_sample_(info));
```

Comment:
I think this logic is not right (it is also wrong in main), because you may finish the loop and take the last sample even if it was rejected.